### PR TITLE
Donot merge DR with different exportTo

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Previously we merge VirtualServices  resides in a namespace even they exportTo different namespaces. This is not right, #52519 reported a inconsistency with this bug.

Now with this patch, it will check exportTo, not merge them until exportTo equals.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@howardjohn @ramaraochavali to have a look and give some feadbacks

File: pilot/pkg/model/destination_rule.go
Comment (MorrisLaw) [inline pilot/pkg/model/destination_rule.go:80]: What's the intention with this logic? It seems to assume that `addRuleToProcessedDestRules` won't be set to `false` if not `true`. Or is this done for readability reasons; to make it clear that it's being set to false here?

If so. Logically, is this the same as the following?
```go
if bothWithoutSelector || (rule.GetWorkloadSelector() != nil && selectorsMatch) && !addRuleToProcessedDestRules {
    addRuleToProcessedDestRules = false
} else {
    addRuleToProcessedDestRules = true
}
```

Comment (hzxuzhonghu) [reply pilot/pkg/model/destination_rule.go:80]: Yes, equivalent

**Comment by howardjohn** [comment]
cc @keithmattix

**Review by keithmattix — COMMENTED** [review]
In my mind this is effectively a precedence question: does the oldest resource heuristic take place first or does exportTo preempt it. If we were starting from scratch, either feels reasonable; honestly, I lean slightly in favor of exportTo having precedence. However, I think this is is still a breaking change and at minimum needs to be behind a compatibility version. I also feel like I need more clarity on the use-case this is enabling as well

**Comment by hzxuzhonghu** [comment]
>does the oldest resource heuristic take place first or does exportTo preempt it

I donot think they are correlated, definitely oldest take preference if the exportTo satisfy. If the oldest doesnot satifsy, then find the newer one


>I think this is is still a breaking change and at minimum needs to be behind a compatibility version

No, this is not a breaking change. Unless i missed something

**Comment by hzxuzhonghu** [comment]
FIx https://github.com/istio/istio/issues/49409 as well

**Comment by hzxuzhonghu** [comment]
@keithmattix

**Comment by hzxuzhonghu** [comment]
ping @keithmattix

**Comment by hzxuzhonghu** [comment]
ping @howardjohn , seems @keithmattix not get this notification

**Comment by hzxuzhonghu** [comment]
ping @howardjohn  @keithmattix how could make the bug fix land down

File: pilot/pkg/model/destination_rule.go
Comment (howardjohn) [inline pilot/pkg/model/destination_rule.go:58]: nit: can use SupersetOf, it doesn't compute a new Set that we immediately GC

File: pilot/pkg/model/destination_rule.go
Comment (howardjohn) [inline pilot/pkg/model/destination_rule.go:53]: Can we add a feature flag? Given all the breakages here, don't feel confident we can make changes like this safely.

Should be pretty easy to do

**Comment by keithmattix** [comment]
I think I would feel comfortable approving if this is behind a feature flag. As @howardjohn mentioned, we've had a lot of breakage in this area of the code lately, and I'm conscious of minimizing risk

File: pilot/pkg/model/destination_rule.go
Comment (hzxuzhonghu) [reply pilot/pkg/model/destination_rule.go:58]: No ,we cannot use superset here. Say if a export to X, while b export export Y.

File: releasenotes/notes/52519.yaml
Comment (howardjohn) [inline releasenotes/notes/52519.yaml:7]: We should add ENABLE_ENHANCED_DESTINATIONRULE_MERGE=false to the compatibility versions so users using those will get the old behavior.

File: pilot/pkg/features/experimental.go
Comment (howardjohn) [inline pilot/pkg/features/experimental.go:201]: This doesn't seem to fully apply, it looks like we only check at some points. I see the same behavior with and without the flag following the reproducer case in https://github.com/istio/istio/issues/52519

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: pilot/pkg/features/experimental.go
Comment (hzxuzhonghu) [reply pilot/pkg/features/experimental.go:201]: Weired, let me have a check with that, maybe later

**Comment by hzxuzhonghu** [comment]
@howardjohn The original issue is on 1.20. #48312 has fixed a case

With this pr feature enabled : 

```
k exec -ti  istiod-6fc7bb65bd-bmpt6      -n istio-system -- curl 127.0.0.1:8080/debug/sidecarz?proxyID=sleep |jq |grep dest -n30
1-{
2:  "destinationRules": {
3-    "httpbin.mesh.global": [
4-      {
5-        "exportTo": {
6-          "default": {},
7-          "test": {}
8-        },
9-        "from": [
10-          {
11-            "Namespace": "test",
12-            "Name": "httpbin-dr"
13-          }
14-        ],
15-        "rule": {
16-          "type": {

```

With this feature disabled

```
k exec -ti  istiod-84f5757589-vf5qf    -n istio-system -- curl 127.0.0.1:8080/debug/sidecarz?proxyID=sleep |jq |grep dest -n30
1-{
2:  "destinationRules": {
3-    "httpbin.mesh.global": [
4-      {
5-        "exportTo": {
6-          "default": {},
7-          "test": {}
8-        },
9-        "from": [
10-          {
11-            "Namespace": "test",
12-            "Name": "httpbin-dr"
13-          },
14-          {
15-            "Namespace": "test",
16-            "Name": "httpbin-ingress-dr"
17-          }
18-        ],
19-        "rule": {

```

**Comment by hzxuzhonghu** [comment]
@howardjohn Can you run the latest code, and have a compare with different exportTo by looking up `/debug/sidecarz`

File: releasenotes/notes/52519.yaml
Comment (ramaraochavali) [inline releasenotes/notes/52519.yaml:9]: The flag name is incorrect

**Review by ramaraochavali — COMMENTED** [review]
left some minor comments and will leave it for @howardjohn to take a look

**Review by howardjohn — COMMENTED** [review]
Overall LGTM, I agree with Rama's comments though

**Review by howardjohn — APPROVED** [review]
Don't want to block this on timezones so I put a hold but since its trivial changes only remaining LGTM

File: pilot/pkg/model/service.go
Comment (hzxuzhonghu) [reply pilot/pkg/model/service.go:309]: Yes, unused

File: releasenotes/notes/52519.yaml
Comment (hzxuzhonghu) [reply releasenotes/notes/52519.yaml:9]: ACK

File: pilot/pkg/features/experimental.go
Comment (hzxuzhonghu) [reply pilot/pkg/features/experimental.go:204]: ACK

File: pilot/pkg/model/push_context.go
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:303]: SidecarScope.MarshalJSON use this, and it it used by /debug/sidecarz

**Pull request merged**
